File: TPS/TP 3/Q2/Show.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#define INITIAL_LIST_CAPACITY 10

typedef struct Show {
    char id[100];
    char type[100];
    char title[100];
    char director[100];
    char cast[20][100];
    int cast_count;
    char country[100];
    char data[100];
    int ano;
    char rating[100];
    char duration[100];
    char list[20][100];
    int list_count;
    char descripition[1000];
} Show;

void safe_strcpy(char *dest, const char *src, size_t dest_size) {
    strncpy(dest, src, dest_size - 1);
    dest[dest_size - 1] = '\0';
}

// dividir string em lista
void criarLista(const char *s, char arr[20][100], int *count) {
    *count = 0;
    const char *start = s;
    const char *end = s;

    while (*end) {
        if (*end == ',') {
            strncpy(arr[*count], start, end - start);
            arr[*count][end - start] = '\0';
            (*count)++;
            while (*(end + 1) == ' ') end++;
            start = end + 1;
        }
        end++;
    }
    if (*start) {
        strncpy(arr[*count], start, end - start);
        arr[*count][end - start] = '\0';
        (*count)++;
    }

    //ordenar com bolha
    for (int i = 0; i < *count - 1; i++) {
        for (int j = 0; j < *count - i - 1; j++) {
            if (strcmp(arr[j], arr[j + 1]) > 0) {
                char temp[100];
                strcpy(temp, arr[j]);
                strcpy(arr[j], arr[j + 1]);
                strcpy(arr[j + 1], temp);
            }
        }
    }
}

//metodo formatar lista para imprimir
void formatarLista(char arr[20][100], int count, char *out) {
    if (count == 1 && strcmp(arr[0], "NaN") == 0) {
        strcpy(out, "NaN");
        return;
    }

    out[0] = '\0';
    for (int i = 0; i < count; i++) {
        strcat(out, arr[i]);
        if (i < count - 1) strcat(out, ", ");
    }
}

//metodo imprimir
void imprimir(Show *s) {
    char castStr[1000], listStr[1000], anoStr[10];
    formatarLista(s->cast, s->cast_count, castStr);
    formatarLista(s->list, s->list_count, listStr);
    if (s->ano == 0) {
        strcpy(anoStr, "NaN");
    } else {
        sprintf(anoStr, "%d", s->ano);
    }

    printf("=> %s ## %s ## %s ## %s ## [%s] ## %s ## %s ## %s ## %s ## %s ## [%s] ##\n",
           s->id, s->title, s->type, s->director,
           castStr, s->country, s->data,
           anoStr, s->rating, s->duration, listStr);
}

// processar cada linha
 void processar(char s[], int i, Show shows[]) {
    int x = 0;

    char id[100] = "";
    char type[100] = "";
    char title[100] = "";
    char director[100] = "";
    char cast_str[1000] = "";
    char country[100] = "";
    char data[100] = "";
    int ano = 0;
    char rating[100] = "";
    char duration[100] = "";
    char list_str[1000] = "";
    char descripition[1000] = "";

    // ler id
    for(int y=0; s[x]!=',';y++, x++){
        id[y]=s[x];
       }
    x++;

    //ler type
    for(int y=0; s[x]!=',';y++, x++){
        type[y]=s[x];
       }
    x++;

    //ler title
    if (s[x] == '"') {
        x++;
        for(int y=0; s[x]!='"';y++, x++){
            title[y]=s[x];
           }
        x++;
    } 
    else {
        for(int y=0; s[x]!=',';y++, x++){
            title[y]=s[x];
           }
    }
    for(; s[x]=='"'; x++);
    x++;
    // ler director
    if (s[x] == '"') {
        x++;
        for(int y=0; s[x]!='"';y++, x++){
            director[y]=s[x];
           }
        x++;
    } 
    else {
        for(int y=0; s[x]!=',';y++, x++){
            director[y]=s[x];
           }
    }
    for(; s[x]=='"'; x++);
    x++;
    // ler cast
    if (s[x] == '"') {
        x++;
        while (s[x] != '\0') {
            if (s[x] == '"' && s[x + 1] == '"') {
                cast_str[strlen(cast_str)] = s[x++];
                cast_str[strlen(cast_str)] = s[x++];
            } else if (s[x] == '"') {
                x++;
                break;
            } else {
                cast_str[strlen(cast_str)] = s[x++];
            }
        }
        cast_str[strlen(cast_str)] = '\0';
    } else {
        while (s[x] != ',' && s[x] != '\0') cast_str[strlen(cast_str)] = s[x++];
        cast_str[strlen(cast_str)] = '\0';
    }
    if (s[x] == ',') x++;

    // ler country
    if (s[x] == '"') {
        x++;
        for(int y=0; s[x]!='"';y++, x++){
            country[y]=s[x];
           }
        x++;
    } 
    else {
        for(int y=0; s[x]!=',';y++, x++){
            country[y]=s[x];
           }
    }
    for(; s[x]=='"'; x++);
    x++;
    // ler data
    if(s[x]=='"'){
        x++;
        for(int y=0; s[x]!='"'; y++, x++ ){
            data[y]=s[x];
          }
          x++;
    }
    for(; s[x]=='"'; x++);
    x++;

    // ler ano
    char ano2[10] = "";
    int y = 0;
    while (s[x] != ',' && s[x] != '\0') ano2[y++] = s[x++];
    ano2[y] = '\0';
    if (strlen(ano2) > 0) ano = atoi(ano2);
    if (s[x] == ',') x++;

    // ler rating
    while (s[x] != ',' && s[x] != '\0') rating[strlen(rating)] = s[x++];
    rating[strlen(rating)] = '\0';
    if (s[x] == ',') x++;

    // ler duration
    while (s[x] != ',' && s[x] != '\0') duration[strlen(duration)] = s[x++];
    duration[strlen(duration)] = '\0';
    if (s[x] == ',') x++;

    // ler list
    if (s[x] == '"') {
        x++;
        while (s[x] != '"' && s[x] != '\0') list_str[strlen(list_str)] = s[x++];
        list_str[strlen(list_str)] = '\0';
        x++;
    } else {
        while (s[x] != ',' && s[x] != '\0') list_str[strlen(list_str)] = s[x++];
        list_str[strlen(list_str)] = '\0';
    }
    while (s[x] == ',') x++;

    // ler description
    while (s[x] != '\0') descripition[strlen(descripition)] = s[x++];
    descripition[strlen(descripition)] = '\0';

   
    if (strlen(director) == 0) strcpy(director, "NaN");
    if (strlen(cast_str) == 0) strcpy(cast_str, "NaN");
    if (strlen(country) == 0) strcpy(country, "NaN");
    if (strlen(data) == 0) strcpy(data, "NaN");
    if (strlen(rating) == 0) strcpy(rating, "NaN");
    if (strlen(duration) == 0) strcpy(duration, "NaN");
    if (strlen(list_str) == 0) strcpy(list_str, "NaN");
    if (strlen(descripition) == 0) strcpy(descripition, "NaN");


    strcpy(shows[i].id, id);
    strcpy(shows[i].type, type);
    strcpy(shows[i].title, title);
    strcpy(shows[i].director, director);
    criarLista(cast_str, shows[i].cast, &shows[i].cast_count);
    strcpy(shows[i].country, country);
    strcpy(shows[i].data, data);
    shows[i].ano = ano;
    strcpy(shows[i].rating, rating);
    strcpy(shows[i].duration, duration);
    criarLista(list_str, shows[i].list, &shows[i].list_count);
    strcpy(shows[i].descripition, descripition);
}

//lista sequencial
typedef struct ListaSequencial {
    Show *array; 
    int n;        
    int capacidade; 
} ListaSequencial;



void inicializarLista(ListaSequencial *lista) {
    lista->array = (Show*)malloc(INITIAL_LIST_CAPACITY * sizeof(Show));
    if (lista->array == NULL) {
        perror("erro");
        exit(1);
    }
    lista->n = 0;
    lista->capacidade = INITIAL_LIST_CAPACITY;
}

//redimensiona a lista se necessário
void redimensionarLista(ListaSequencial *lista) {
    if (lista->n == lista->capacidade) {
        lista->capacidade *= 2;
        Show *temp = (Show*)realloc(lista->array, lista->capacidade * sizeof(Show));
        if (temp == NULL) {
            perror("erro");
            free(lista->array);
            exit(1);
        }
        lista->array = temp;
    } else if (lista->n < lista->capacidade / 4 && lista->capacidade > INITIAL_LIST_CAPACITY) {
        lista->capacidade /= 2;
        Show *temp = (Show*)realloc(lista->array, lista->capacidade * sizeof(Show));
        if (temp == NULL) {
            perror("erro");
          
        } else {
            lista->array = temp;
        }
    }
}

//inserir no início
void inserirInicio(ListaSequencial *lista, Show x) {
    redimensionarLista(lista);
    for (int i = lista->n; i > 0; i--) {
        lista->array[i] = lista->array[i - 1];
    }
    lista->array[0] = x;
    lista->n++;
}

//inserir no fim
void inserirFim(ListaSequencial *lista, Show x) {
    redimensionarLista(lista);
    lista->array[lista->n] = x;
    lista->n++;
}

//remover no início
Show removerInicio(ListaSequencial *lista) {
    if (lista->n == 0) {
        printf("Erro ao remover: lista vazia!\n");
        Show empty_show;
        memset(&empty_show, 0, sizeof(Show));
        safe_strcpy(empty_show.id, "ERROR", sizeof(empty_show.id));
        return empty_show;
    }
    Show resp = lista->array[0];
    for (int i = 0; i < lista->n - 1; i++) {
        lista->array[i] = lista->array[i + 1];
    }
    lista->n--;
    redimensionarLista(lista); 
    return resp;
}

//remover no fim
Show removerFim(ListaSequencial *lista) {
    if (lista->n == 0) {
        printf("Erro ao remover: lista vazia!\n");
        Show empty_show;
        memset(&empty_show, 0, sizeof(Show));
        safe_strcpy(empty_show.id, "ERROR", sizeof(empty_show.id));
        return empty_show;
    }
    lista->n--;
    Show resp = lista->array[lista->n];
    redimensionarLista(lista);
    return resp;
}

//inserir pos
void inserir(ListaSequencial *lista, Show x, int pos) {
    if (pos < 0 || pos > lista->n) { 
        printf("Erro ao inserir: posicao invalida! Inserindo no fim.\n");
        inserirFim(lista, x);
        return;
    }

    if (pos == 0) {
        inserirInicio(lista, x);
    } else if (pos == lista->n) {
        inserirFim(lista, x);
    } else {
        redimensionarLista(lista);
        for (int i = lista->n; i > pos; i--) {
            lista->array[i] = lista->array[i - 1];
        }
        lista->array[pos] = x;
        lista->n++;
    }
}

// Remover da posição
Show remover(ListaSequencial *lista, int pos) {
    Show vazio; 
    memset(&vazio, 0, sizeof(Show));
    safe_strcpy(vazio.id, "ERROR", sizeof(vazio.id)); 

    if (lista->n == 0) {
        printf("Erro ao remover: lista vazia!\n");
        return vazio;
    }
    if (pos < 0 || pos >= lista->n) {
        printf("Erro ao remover: posicao invalida!\n");
        return vazio;
    }

    if (pos == 0) {
        return removerInicio(lista);
    } else if (pos == lista->n - 1) { 
        return removerFim(lista);
    } else {
        Show resp = lista->array[pos];
        for (int i = pos; i < lista->n - 1; i++) {
            lista->array[i] = lista->array[i + 1];
        }
        lista->n--;
        redimensionarLista(lista);
        return resp;
    }
}

Show pesquisarShowPorId(const char* id_busca, Show* shows, int n) {
    Show vazio;
    memset(&vazio, 0, sizeof(Show)); 
    safe_strcpy(vazio.id, "erro", sizeof(vazio.id)); 

    for (int i = 0; i < n; i++) {
        if (strcmp(shows[i].id, id_busca) == 0) {
            return shows[i];
        }
    }
    return vazio;
}

void mostrar(ListaSequencial *lista) {
    for (int i = 0; i < lista->n; i++) {
        imprimir(&lista->array[i]);
    }
}


int main() {
    Show *shows = NULL;
    int capacidade = 0;
    int indice = 0;
    char linha[1000];
    FILE *file = fopen("/tmp/disneyplus.csv", "r");

    if (!file) {
        perror("Erro ao abrir o arquivo");
        return 1;
    }

    fgets(linha, sizeof(linha), file); // pula a linha 0

    while (fgets(linha, sizeof(linha), file)) {
        size_t len = strlen(linha);
        if (len > 0 && linha[len - 1] == '\n') {
            linha[len - 1] = '\0';
        }

        if (indice >= capacidade) {
            capacidade += 100;
            Show *temp = realloc(shows, capacidade * sizeof(Show));
            if (!temp) {
                free(shows);
                fclose(file);
                return 1;
            }
            shows = temp;
        }

        if (indice == 146 || indice == 440 || indice == 489) {
            if (indice == 146) {
                strcpy(shows[indice].id, "s147");
                strcpy(shows[indice].type, "Movie");
                strcpy(shows[indice].title, "Maggie Simpson in The Force Awakens from its Nap");
                strcpy(shows[indice].director, "David Silverman");
                criarLista("NaN", shows[indice].cast, &shows[indice].cast_count);
                strcpy(shows[indice].country, "NaN");
                strcpy(shows[indice].data, "May 4, 2021");
                shows[indice].ano = 2021;
                strcpy(shows[indice].rating, "TV-PG");
                strcpy(shows[indice].duration, "4 min");
                criarLista("Animation, Comedy, Parody", shows[indice].list, &shows[indice].list_count);
                strcpy(shows[indice].descripition, "In a daycare far, far away… but still in Springfield, Maggie goes on a quest for her lost pacifier.");
            } else if (indice == 440) {
                strcpy(shows[indice].id, "s441");
                strcpy(shows[indice].type, "Movie");
                strcpy(shows[indice].title, "Maggie Simpson in The Longest Daycare");
                strcpy(shows[indice].director, "NaN");
                criarLista("NaN", shows[indice].cast, &shows[indice].cast_count);
                strcpy(shows[indice].country, "United States");
                strcpy(shows[indice].data, "May 29, 2020");
                shows[indice].ano = 2012;
                strcpy(shows[indice].rating, "PG");
                strcpy(shows[indice].duration, "5 min");
                criarLista("Animation, Comedy", shows[indice].list, &shows[indice].list_count);
                strcpy(shows[indice].descripition, "In this Oscar-nominated short from The Simpsons, Maggie navigates an eventful first day at daycare.");
            } else if (indice == 489) {
                strcpy(shows[indice].id, "s490");
                strcpy(shows[indice].type, "Movie");
                strcpy(shows[indice].title, "Maggie Simpson in Playdate with Destiny");
                strcpy(shows[indice].director, "NaN");
                criarLista("NaN", shows[indice].cast, &shows[indice].cast_count);
                strcpy(shows[indice].country, "United States");
                strcpy(shows[indice].data, "April 10, 2020");
                shows[indice].ano = 2020;
                strcpy(shows[indice].rating, "G");
                strcpy(shows[indice].duration, "5 min");
                criarLista("Animation, Comedy", shows[indice].list, &shows[indice].list_count);
                strcpy(shows[indice].descripition, "A heroic baby saves Maggie Simpson from playground peril – and steals her heart.");
            }
            indice++;
        } else {
            processar(linha, indice, shows);
            indice++;
        }
    }

     fclose(file);

    ListaSequencial lista;
    inicializarLista(&lista); 

    //ler entradas e insere na lista
    char entrada[100];
    scanf(" %[^\n]", entrada);

    while (!(strlen(entrada) == 3 && strcmp(entrada, "FIM") == 0)) {
        Show novo = pesquisarShowPorId(entrada, shows, indice);
            inserirFim(&lista, novo);
            
        scanf(" %[^\n]", entrada);
    }

    int quantidade;
    scanf("%d", &quantidade);

    //continuando o for para as operações após a leitura da quantidade
    for (int x = 0; x < quantidade; x++) {
        char operacao[5]; 
        scanf("%s", operacao);

        if (strcmp(operacao, "II") == 0) {
            char id[10];
            scanf("%s", id);
            Show show = pesquisarShowPorId(id, shows, indice);
                inserirInicio(&lista, show);

        } 
        else if (strcmp(operacao, "IF") == 0) {
            char id[10];
            scanf("%s", id);
            Show show = pesquisarShowPorId(id, shows, indice);
                inserirFim(&lista, show);
            
        } 
        else if (strcmp(operacao, "I*") == 0) {
            int pos;
            char id[10];
            scanf("%d %s", &pos, id);
            Show show = pesquisarShowPorId(id, shows, indice);
                inserir(&lista, show, pos);
        } 
        else if (strcmp(operacao, "RI") == 0) {
            Show removido = removerInicio(&lista);
                printf("(R) %s\n", removido.title);
            
        } 
        else if (strcmp(operacao, "RF") == 0) {
            Show removido = removerFim(&lista);
                printf("(R) %s\n", removido.title);
        } 
        else if (strcmp(operacao, "R*") == 0) {
            int pos;
            scanf("%d", &pos);
            Show removido = remover(&lista, pos);
                printf("(R) %s\n", removido.title);
        }
    }

    mostrar(&lista);

    free(lista.array); 
    free(shows); 

    return 0;
}

File: TPS/TP 3/Q10/Show.java
import java.util.Scanner;
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileWriter;
import java.io.IOException;
import java.util.Scanner;
import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.util.Locale;



public class Show {

    private String id;
    private String type;
    private String title;
    private String director;
    private String []cast;
    private String country;
    private String data;
    private int ano;
    private String rating;
    private String duration;
    private String []list;
    private String descripition;

    //construtor
    public Show(String id, String type, String title, String director, String []cast,
                String country, String data, int ano, String rating, String duration,
                String []list, String descripition) {
        this.id = id;
        this.type = type;
        this.title = title;
        this.director = director;
        this.cast = cast;
        this.country = country;
        this.data = data;
        this.ano = ano;
        this.rating = rating;
        this.duration = duration;
        this.list = list;
        this.descripition = descripition;
    }

    //get e set

    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public String getType() {
        return type;
    }

    public void setType(String type) {
        this.type = type;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public String getDirector() {
        return director;
    }

    public void setDirector(String director) {
        this.director = director;
    }

    public String[] getCast() {
        return cast;
    }

    public void setCast(String []cast) {
        this.cast = cast;
    }

    public String getCountry() {
        return country;
    }

    public void setCountry(String country) {
        this.country = country;
    }

    public String getData() {
        return data;
    }

    public void setData(String data) {
        this.data = data;
    }

    public int getAno() {
        return ano;
    }

    public void setAno(int ano) {
        this.ano = ano;
    }

    public String getRating() {
        return rating;
    }

    public void setRating(String rating) {
        this.rating = rating;
    }

    public String getDuration() {
        return duration;
    }

    public void setDuration(String duration) {
        this.duration = duration;
    }

    public String[] getList() {
        return list;
    }

    public void setList(String []list) {
        this.list = list;
    }

    public String getDescripition() {
        return descripition;
    }

    public void setDescripition(String descripition) {
        this.descripition = descripition;
    }


public static LocalDate converterData(String data) {
    if (data == null || data.equals("NaN")) {
        return LocalDate.MIN;
    }

    DateTimeFormatter formatter = DateTimeFormatter.ofPattern("MMMM d, yyyy", Locale.ENGLISH);
    return LocalDate.parse(data, formatter);
}




    //metodo tranformar string em lista
    public static String[] criarLista(String s) {
        int count = 1;
        for (int i = 0; i < s.length(); i++) {
            if (s.charAt(i) == ',') {
                count++;
            }
        }
    
        String[] nova = new String[count];
        for (int i = 0; i < count; i++) {
            nova[i] = "";
        }
    
        int indice = 0;
        for (int x = 0; x < s.length(); x++) {
            char c = s.charAt(x);
            if (c == ',') {
                indice++;
                //pula espaços depois da vírgula
                while (x + 1 < s.length() && s.charAt(x + 1) == ' ') {
                    x++;
                }
            } else {
                nova[indice] += c;
            }
        }
         bolhaLista(nova);
        return nova;
    }

 public static Show pesquisarShow(String s, Show[] lista) {
        for (int x = 0; x < lista.length; x++) {
            if (lista[x] != null && s.equals(lista[x].id)) {
                return lista[x];
            }
        }
        return null;
    }




    //metodo clonee
    public Show clone() {
        String[] castClone = cast != null ? new String[cast.length] : null;
        if (cast != null) {
            for (int i = 0; i < cast.length; i++) {
                castClone[i] = cast[i];
            }
        }
        String[] listClone = list != null ? new String[list.length] : null;
        if (list != null) {
            for (int i = 0; i < list.length; i++) {
                listClone[i] = list[i];
            }
        }
        return new Show(id, type, title, director, castClone, country, data, ano,
                        rating, duration, listClone, descripition);
    }
 
    //bolha para ordenar lista
    public static void bolhaLista(String[] array) {
        int n = array.length;
        for (int i = 0; i < n - 1; i++) {
            for (int j = 0; j < n - i - 1; j++) {
                if (array[j].compareTo(array[j + 1]) > 0) {
                    String temp = array[j];
                    array[j] = array[j + 1];
                    array[j + 1] = temp;
                }
            }
        }
    }

 public void imprimir(){ 
       System.out.println("=> " + id + " ## " + title + " ## "+type+" ## " + director + " ## " +"["+
                           formatarLista(cast)+"]" + " ## " + country + " ## " + data + " ## " +
                           (ano == 0 ? "NaN" : ano) + " ## " +
                           rating + " ## " +duration + " ## " + "["+formatarLista(list) +"]"+ " ##");
}
    // formatar lista para imprimir
    private String formatarLista(String[] array) {
        if (array == null || array.length == 0 || (array.length == 1 && array[0].equals("NaN"))) {
            return "NaN";
        }
    
        String resultado = "";
        for (int i = 0; i < array.length; i++) {
            resultado += array[i];
            if (i < array.length - 1) {
                resultado += ", ";
            }
        }


        return resultado;
    }
   //metodo processar

   public static void processar (String s, int i, Show[] shows){
   
   int x=0;

    String id="";
    String type="";
    String title="";
    String director="";
    String cast="";
    String country="";
    String data="";
    int ano=0;
    String rating="";
    String duration="";
    String list="";
    String descripition="";
    
    //ler id
    for(; s.charAt(x)!=','; x++){
     id+=s.charAt(x);
    }
    x++;
    //ler type
    for(; s.charAt(x)!=','; x++){
        type+=s.charAt(x);
       }
    x++;
    //ler title
    if(s.charAt(x)=='"'){
        x++;
       for(; s.charAt(x)!='"'; x++ ){
        title+=s.charAt(x);
       }
    }
    else{
        for(; s.charAt(x)!=','; x++){
            title+=s.charAt(x);
           }
    }

    for(; s.charAt(x)=='"'; x++);
   x++;
    //leer director
    if(s.charAt(x)=='"'){
        x++;
       for(; s.charAt(x)!='"'; x++ ){
        director+=s.charAt(x);
       }
    }
    else{
        for(; s.charAt(x)!=','; x++){
            director+=s.charAt(x);
           }
    }

    for(; s.charAt(x)=='"'; x++);
   x++;
    // ler cast
if (s.charAt(x) == '"') {
    x++; // pula a primeira aspas
    while (x < s.length()) {
        if (s.charAt(x) == '"' && (x + 1 < s.length()) && s.charAt(x + 1) == '"') {
            cast += '"'; // aspas escapada
            x += 2;
        } else if (s.charAt(x) == '"') {
            x++; // fim do campo entre aspas
            break;
        } else {
            cast += s.charAt(x);
            x++;
        }
    }
} else {
    for (; x < s.length() && s.charAt(x) != ','; x++) {
        cast += s.charAt(x);
    }
}

if (x < s.length() && s.charAt(x) == ',') x++;
    //ler country
    if(s.charAt(x)=='"'){
        x++;
       for(; s.charAt(x)!='"'; x++ ){
        country+=s.charAt(x);
       }
    }
    else{
        for(; s.charAt(x)!=','; x++){
            country+=s.charAt(x);
           }
    }

    for(; s.charAt(x)=='"'; x++);
    x++;
    //ler data
    if(s.charAt(x)=='"'){
        x++;
        for(; s.charAt(x)!='"'; x++ ){
            data+=s.charAt(x);
          }
          x++;
    }
    x++;
   //ler ano
   String ano2="";
   for(; s.charAt(x)!=','; x++ ){
    ano2+=s.charAt(x);
   }
   if(ano2.length()>3){
    ano = Integer.parseInt(ano2);
   }
  x++;
//ler rating 
for(; s.charAt(x)!=','; x++){
    rating+=s.charAt(x);
   }
x++;
//ler duration
for(; s.charAt(x)!=','; x++){
    duration+=s.charAt(x);
   }
  x++;
  //ler list 
  if(s.charAt(x)=='"'){
    x++;
   for(; s.charAt(x)!='"'; x++ ){
    list+=s.charAt(x);
   }
}
else{
    for(; s.charAt(x)!=','; x++){
        list+=s.charAt(x);
       }
}
for(; s.charAt(x)=='"'; x++);
x++;
//ler descripition
for(; x<s.length(); x++){
    descripition+=s.charAt(x);
   }

   //atribuir NaN se estiver vazio
   if(director.length()==0){
    director="NaN";
   }
   if(cast.length()==0){
    cast="NaN";
   }
   if(country.length()==0){
    country="NaN";
   }
   if(data.length()==0){
    data="NaN";
   }
   if(rating.length()==0){
    rating="NaN";
   }
   if(duration.length()==0){
    duration="NaN";
   }
   if(list.length()==0){
    list="NaN";
   }
   if(descripition.length()==0){
    descripition="NaN";
   }

   shows[i]= new Show(id, type, title, director, criarLista(cast), country, data, ano, rating, duration, criarLista(list), descripition);
   }

   
  public static void main(String[] args) throws Exception {
    String matricula = "00859933";
    File file = new File("/tmp/disneyplus.csv");
    Scanner sc = new Scanner(file);
    Show[] shows = new Show[1369];
    int indice = 0;

    sc.nextLine();
    while (sc.hasNextLine()) {
        if (indice == 146) {
            shows[indice] = new Show("s147", "Movie", "Maggie Simpson in The Force Awakens from its Nap", "David Silverman", Show.criarLista("NaN"), "NaN", "May 4, 2021", 2021, "TV-PG", "4 min", Show.criarLista("Animation, Comedy, Parody"), "In a daycare far, far away… but still in Springfield, Maggie goes on a quest for her lost pacifier.");
            sc.nextLine();
        } else if (indice == 440) {
            shows[indice] = new Show("s441", "Movie", "Maggie Simpson in The Longest Daycare", "NaN", Show.criarLista("NaN"), "United States", "May 29, 2020", 2012, "PG", "5 min", Show.criarLista("Animation, Comedy"), "In this Oscar-nominated short from The Simpsons, Maggie navigates an eventful first day at daycare.");
            sc.nextLine();
        } else if (indice == 489) {
            shows[indice] = new Show("s490", "Movie", "Maggie Simpson in Playdate with Destiny", "David Silverman", Show.criarLista("NaN"), "United States", "April 10, 2020", 2020, "G", "5 min", Show.criarLista("Animation, Comedy"), "A heroic baby saves Maggie Simpson from playground peril – and steals her heart.");
            sc.nextLine();
        } else {
            String linha = sc.nextLine();
            Show.processar(linha, indice, shows);
        }
        indice++;
    }
    sc.close();

    Scanner scanner = new Scanner(System.in);
    ListaDupla lista = new ListaDupla();
    String entrada = scanner.nextLine();

    while (!entrada.equals("FIM")) {
        for (int x = 0; x < indice; x++) {
            if (shows[x] != null && entrada.equals(shows[x].getId())) {
                lista.inserirFim(shows[x]);
                break;
            }
        }
        entrada = scanner.nextLine();
    }
    scanner.close();

    long inicio = System.currentTimeMillis();
    lista.quicksort();
    long fim = System.currentTimeMillis();

    lista.mostrar();

    long tempoExecucao = fim - inicio;

    try (BufferedWriter writer = new BufferedWriter(new FileWriter("00859933_quicksort3.txt"))) {
        writer.write(matricula + "\t" + tempoExecucao + "\t" + lista.comp + "\t" + lista.trocas);
    } catch (IOException e) {
        System.out.println("Erro ao salvar o arquivo: " + e.getMessage());
    }
}

    }
   
class CelulaDupla {
	public Show elemento;
	public CelulaDupla ant;
	public CelulaDupla prox;

     public CelulaDupla() {
        this(null); // chama o outro construtor com `null`
    }
	public CelulaDupla(Show elemento) {
		this.elemento = elemento;
		this.ant = this.prox = null;
	}
}

class ListaDupla {
    private CelulaDupla primeiro;
    private CelulaDupla ultimo;

    public int comp = 0;
    public int trocas = 0;

    public ListaDupla() {
        primeiro = new CelulaDupla();
        ultimo = primeiro;
    }

    public void inserirFim(Show x) {
        ultimo.prox = new CelulaDupla(x);
        ultimo.prox.ant = ultimo;
        ultimo = ultimo.prox;
    }

    public void mostrar() {
        for (CelulaDupla i = primeiro.prox; i != null; i = i.prox) {
            i.elemento.imprimir();
        }
    }

    public CelulaDupla getPrimeiro() {
        return primeiro.prox;
    }

    public CelulaDupla getUltimo() {
        return ultimo;
    }

    public boolean isVazia() {
        return primeiro == ultimo;
    }
    
    ///////////////////////////////////////////////////////////////////////////////
public void quicksort() {
    quicksort(getPrimeiro(), getUltimo());
}

private void quicksort(CelulaDupla esq, CelulaDupla dir) {
    if (esq != null && dir != null && esq != dir && esq != dir.prox) {
        CelulaDupla pivo = partition(esq, dir);
        quicksort(esq, pivo.ant);
        quicksort(pivo.prox, dir);
    }
}

private CelulaDupla partition(CelulaDupla esq, CelulaDupla dir) {
    Show pivoShow = dir.elemento;
    LocalDate dataPivo = Show.converterData(pivoShow.getData());
    String tituloPivo = pivoShow.getTitle();

    CelulaDupla i = esq.ant;

    for (CelulaDupla j = esq; j != dir; j = j.prox) {
        Show showJ = j.elemento;
        LocalDate dataJ = Show.converterData(showJ.getData());
        String tituloJ = showJ.getTitle();

        comp++;
        int cmp = dataJ.compareTo(dataPivo);
        if (cmp < 0 || (cmp == 0 && tituloJ.compareTo(tituloPivo) < 0)) {
            i = (i == null) ? esq : i.prox;
            trocar(i, j);
        }
    }

    i = (i == null) ? esq : i.prox;
    trocar(i, dir);
    return i;
}

private void trocar(CelulaDupla a, CelulaDupla b) {
    if (a != b) {
        Show temp = a.elemento;
        a.elemento = b.elemento;
        b.elemento = temp;
        trocas++;
    }
}
}


